### PR TITLE
DOC: Monkey testing with {shinytest2} 

### PR DESCRIPTION
**Pull request by DivadNojnarg**
@schloerke : I just started the vignette. You'll be able to see the progress here.

### CODE REVIEW

**Comment by schloerke** [comment]
Thank you!

**Comment by DivadNojnarg** [comment]
@schloerke : Do the latest API changes require to update the vignette? Thinking about `run_js` vs `execute_js`

File: vignettes/monkey-testing.Rmd [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+The easiest way is to call:
+
+```r
+headless_app$execute_js("
+  (function() {
+    var s = document.createElement('script');
+    s.src = 'https://unpkg.com/gremlins.js';
+    document.body.appendChild(s);
+  })();"
+)
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:107]:
```suggestion
headless_app$run_js("
  let s = document.createElement('script');
  s.src = 'https://unpkg.com/gremlins.js';
  document.body.appendChild(s);
")
```

File: vignettes/monkey-testing.Rmd [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+This creates a `<script>` tag pointing to the right CDN (content delivery network: optimized server to store libraries) and inserts
+it at the end of the body.
+
+To test whether everything worked well, we can dump the DOM, look for the scripts. We should be able to find
+gremlins.js and try to call `typeof window.gremlins`, which should return an object:
+
+```r
+headless_app$get_html("script", outer_html = TRUE)
+headless_app$execute_js("return typeof window.gremlins")
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:118]:
```suggestion
headless_app$get_js("typeof window.gremlins")
```

File: vignettes/monkey-testing.Rmd [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+```r
+headless_app$execute_js("
+  (function() {
+    var s = document.createElement('script');
+    s.src = './gremlins/gremlins.min.js';
+    document.body.appendChild(s);
+  })();"
+)
+plop$get_html("script", outer_html = TRUE)
+plop$execute_js("return typeof window.gremlins")
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:150]:
```suggestion
headless_app$run_js("
  let s = document.createElement('script');
  s.src = './gremlins/gremlins.min.js';
  document.body.appendChild(s);
")
plop$get_html("script", outer_html = TRUE)
plop$get_js("typeof window.gremlins")
```

File: vignettes/monkey-testing.Rmd [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+
+
+### Practice
+
+#### Blind run
+
+In the following, we run the most basic Monkey test configuration:
+
+```r
+headless_app$execute_js("gremlins.createHorde().unleash();")
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:183]:
```suggestion
headless_app$run_js("gremlins.createHorde().unleash();")
```

File: vignettes/monkey-testing.Rmd [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+    
+    gremlins.createHorde({
+      //randomizer: new gremlins.Chance(1234), // repeatable
+      species: [customToucher],
+      mogwais: [gremlins.mogwais.gizmo()]
+    }).unleash().then(() => {
+      console.log('Gremlins test success')
+    });
+  "
+)
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:262]:
```suggestion
headless_app$run_js("
  const customToucher = gremlins.species.toucher({
    // which touch event types will be triggered
    touchTypes: ['gesture'],
    // Touch only if element has class irs-handle single
    canTouch: (element) => element.className === 'irs-handle single',
    log: true,
    maxTouches: 200
  });
    
  gremlins.createHorde({
    //randomizer: new gremlins.Chance(1234), // repeatable
    species: [customToucher],
    mogwais: [gremlins.mogwais.gizmo()]
  }).unleash().then(() => {
    console.log('Gremlins test success')
  });
")
```

**Review by schloerke — COMMENTED** [review]
Small changes to use `$get_js()` or `$run_js()`

File: vignettes/monkey-testing.Rmd [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+
+
+### Local way
+
+If you work in a corporate environment chances are you'll end up in this situation. 
+A solution is then to store and serve a local copy of the gremlin.js script with `shiny::addResourcePath()`.
+Assuming `gremlins.js` are in `inst/js/gremlins.min.js`. Add this to the `app.R` file:
+
+```r
+addResourcePath("gremlins", "inst/js/gremlins.min.js")
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:134]:
```suggestion
shiny::addResourcePath("gremlins", "inst/js/gremlins.min.js")
```

File: vignettes/monkey-testing.Rmd [hunk 7]
```diff
@@ -0,0 +1,279 @@
...
+We can subsequently inject the gremlins in the DOM and checks whether everything worked as expected:
+
+```r
+headless_app$run_js("
+  let s = document.createElement('script');
+  s.src = './gremlins/gremlins.min.js';
+  document.body.appendChild(s);
+")
+plop$get_html("script", outer_html = TRUE)
+plop$get_js("typeof window.gremlins")
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:146]:
```suggestion
headless_app$get_html("script", outer_html = TRUE)
headless_app$get_js("typeof window.gremlins")
```

File: vignettes/monkey-testing.Rmd [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+const horde = gremlins.createHorde();
+horde.unleash();
+```
+
+`createHorde()` accepts many species of gremlins capable of handling various events such as
+clicks, touch, form filling, scrolling, typing, ... described in the gremlins.js [documentation](https://github.com/marmelab/gremlins.js#advanced-usage).
+
+As a side note, we don't recommend using the `scroller` which sometimes [crashes](https://github.com/rstudio/shinytest2/issues/117) the Chrome instance. 
+
+If your plots rely on random elements like `rnorm`, it is best practice to setup a seed.
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:168]:
```suggestion
If your plots rely on random elements like `rnorm`, it is best practice to setup a `seed`. 
```

File: vignettes/monkey-testing.Rmd [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+horde.unleash();
+```
+
+`createHorde()` accepts many species of gremlins capable of handling various events such as
+clicks, touch, form filling, scrolling, typing, ... described in the gremlins.js [documentation](https://github.com/marmelab/gremlins.js#advanced-usage).
+
+As a side note, we don't recommend using the `scroller` which sometimes [crashes](https://github.com/rstudio/shinytest2/issues/117) the Chrome instance. 
+
+If your plots rely on random elements like `rnorm`, it is best practice to setup a seed. 
+By default, all species will attack in random order with a delay of 10 ms between each event. You can also control the attack strategy to fine tune the global behavior. Finally, if you want more control over what gremlins specie should be doing, you can define a custom specie.
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:169]:
```suggestion
By default, all species will attack in random order with a delay of 10 ms between each event. You can also control the attack strategy to fine tune the global behavior. Finally, if you want more control over what your gremlins species should be doing, you can define a custom species.
```

File: vignettes/monkey-testing.Rmd [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+#### Optimized testing
+
+Does it really makes sense to invoke all species given we only have a slider? Absolutely not!
+In the following code, we only want to test the slider input.
+
+```{r echo=FALSE, out.width='100%', fig.align='center'}
+knitr::include_graphics("figures/gremlins-slider-handle.png")
+```
+
+The best specie to perform this task is the `toucher`, which is able to randomly move the slider input. The documentation specifies many events such as tap, doubletap, gesture and multitouch. It seems more relevant to apply only gesture, consisting of dragging the slider on the x-axis (note: gesture also consider the y-axis which does not make sense for the slider, which is an gremlins.js limitation. This however won't prevent our test from running). Also, we target the slider handle by its class `irs-handle single` to ensure we don't touch any other element and increase the number of maximum touch from 2 to 200. Not how precious is the HTML inspector, which allows to seamlessly inspect and extract any class or id. The __log__ parameter enables log in the JavaScript console:
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:197]:
```suggestion
The best species to perform this task is the `toucher`, which is able to randomly move the slider input. The documentation specifies many events such as tap, doubletap, gesture and multitouch. It seems more relevant to apply only gesture, consisting of dragging the slider on the x-axis (note: gesture also consider the y-axis which does not make sense for the slider, which is a gremlins.js limitation. This however won't prevent our test from running). Also, we target the slider handle by its class `irs-handle single` to ensure we don't touch any other element and increase the number of maximum touch from 2 to 200. Note how precious is the HTML inspector, which allows to seamlessly inspect and extract any class or id. The __log__ parameter enables logging in the JavaScript console:
```

File: vignettes/monkey-testing.Rmd [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+    //randomizer: new gremlins.Chance(1234), // repeatable
+    species: [customToucher],
+    mogwais: [gremlins.mogwais.gizmo()]
+  }).unleash().then(() => {
+    console.log('Gremlins test success')
+  });
+")
+
+Sys.sleep(3)
+headless_app$screenshot("gremlins.png")
```
Comment (schloerke) [inline vignettes/monkey-testing.Rmd:260]:
```suggestion
headless_app$get_screenshot("gremlins.png")
```

**Review by schloerke — APPROVED** [review]

**Comment by schloerke** [comment]
Thank you so much @DivadNojnarg !

**Pull request merged**